Next version update (March 2020)

rstacruz opened the pull request:
> **WIP:** While this is a work-in-progress, therelease is available under the `next` tag - try it out with `npm install nprogress@next` 😉  [(draft release notes)](https://github.com/rstacruz/nprogress/releases)

This updates NProgress to keep it up-to-date with 2020 standards.

## Major updates

- **Distirbute an ESM build.** NProgress is currently a UMD distribution (ie, CommonJS and browser).

- **Add CSS variables support.** It should be easier to change the color.

- **Add Sass support.** Same goal as above for those who prefer Sass.

- **Types and documentation.** This should hopefully make the experience better for TypeScript and VSCode users.

## Build process updates

- [ ] Use [`jest`](https://jestjs.io) for tests (rather than Mocha and Chai).
- [x] Use [`prettier`](https://prettier.io) for code formatting.
- [x] Use [`microbundle`](https://npm.im/microbundle) to distribute UMD, ESM, and CommonJS builds.
- [x] Use [`yarn`](https://yarn.pm) for packaging.
- [ ] Use [`np`](https://npm.im/np) for publishing.

## CI updates

- [x] Move from Travis CI to GitHub actions.

## Deprecations

- [x] Remove Bower support.
- [x] Remove [component.json](https://github.com/componentjs/spec) support.
- [x] Remove jspm support.

## Documentation update

- [ ] Update jsdoc comments (with types). This should make the experience better for TypeScript and VSCode users.
- [x] Move the website to the gh-pages branch.
- [x] Update the website a bit.
- [ ] Update README.
- [x] Update CONTRIBUTING.md for notes on maintenance.
- [x] Migrate changelog to GitHub releases

## CSS updates

- [x] Add CSS variables support <!-- https://github.com/rstacruz/nprogress/pull/219/files -->
- [x] Add Sass support

rstacruz committed:
Remove bower.json and component.json

rstacruz committed:
Convert to esm

rstacruz committed:
Replace test suite with Jest

rstacruz committed:
Remove jspm support

rstacruz committed:
Use babel

rstacruz committed:
Configure microbundle; put css in css/

rstacruz committed:
Add GitHub actions support, remove Travis

rstacruz committed:
Add some jsdoc annotations

anselal commented:
Nicely done so far, keep up the good work.
FYI, your "little" project is used widely, so you done right to update the js file to keep up with 2020 standards as well adding `sass` support so it can be used with new JS frameworks !!!

anselal committed:
:star: add(css_varriables): Add variables for bar height and color

rstacruz committed:
Change notes to CONTRIBUTING.md

rstacruz committed:
Remove the website from the repo

rstacruz committed:
Rename Markdown files

anselal commented:
I created a PR for css variables

rstacruz committed:
Refactor CSS

rstacruz committed:
Add support for --nprogress-height variables

rstacruz committed:
Update CHANGELOG

rstacruz committed:
Merge branch 'rstacruz-2020-04-04' into rstacruz-2020-04-04

rstacruz committed:
Merge branch 'rstacruz-2020-04-04' into rstacruz-2020-04-04

rstacruz committed:
Merge pull request #219 from anselal/rstacruz-2020-04-04

rstacruz committed:
Merge remote-tracking branch 'origin/master' into rstacruz-2020-04-04

rstacruz committed:
Add Sass support

rstacruz committed:
Add prettier formatting

rstacruz committed:
Add package.json keywords

rstacruz committed:
Update changelog

rstacruz committed:
v1.0.0-0

rstacruz committed:
Set registry for pushing

rstacruz committed:
v1.0.0-1

rstacruz committed:
Remove CHANGELOG.md

rstacruz committed:
Remove NProrgress.promise()

rstacruz committed:
Add more type annotations

rstacruz committed:
Upgrade jest

rstacruz committed:
Refactor much of the README

- Change unpkg to cdn.jsdelivr.net
- Remove customization section
- Add jQuery integration notes

anselal commented on README.md:
````diff
@@ -1,225 +1,264 @@
 
-Also available via [unpkg] CDN:
+```css
+@import "~nprogress/css/nprogress.css";
````
You could also add the scss import so the users will know it exists ;)

rstacruz committed:
Update tests and refactor DOM utilities to dom.js

rstacruz committed:
Extract CSS helper to css.js

rstacruz committed:
Move more utilities to utilities.js

rstacruz committed:
Refactor queue, this, and others

rstacruz committed:
Deprecate .status in favor of  .getPercent()

anselal commented:
Looking good my friend. Keep up the good work !!!

rstacruz committed:
Use fallback CSS variable values

rstacruz committed:
Add demo, update documentation

rstacruz committed:
Update defaults to be a bit more reasonable

rstacruz committed:
Begin rewriting the README

vvo commented:
Just chiming in to say: I just tried `@next` release and it worked perfectly. Only thin I had to change was the CSS import, to `import "nprogress/css/nprogress.css";` Thanks for working on upgrading the library, it's a great library. I recently used it in a Next.js project to show the progress bar, after a delay, on Route en fetch events (https://dev.to/vvo/show-a-top-progress-bar-on-fetch-and-router-events-in-next-js-4df3)

vvo commented:
@rstacruz For what it's worth: I did some more testing and on iOS Safari the behavior of the progress bar is different than before. When start() is called and it takes a "long" (like 3s) time before done(), the bar progress is not smooth. It jumps back and forth constantly like it's freezing.

I was able to reproduce the behavior in a very simple example (using `@next`):
- live preview: https://wvwgj.csb.app/
- code: https://codesandbox.io/s/beautiful-solomon-wvwgj?file=/src/index.js

Demo:
![RPReplay_Final1591869653](https://user-images.githubusercontent.com/123822/84372587-9f7a4780-abdb-11ea-8297-d3694b21c009.gif)

Same CodeSandbox but **using `@latest` instead of `@next`**:
- live preview: https://m5srm.csb.app/
- code: https://codesandbox.io/s/intelligent-agnesi-m5srm?file=/src/index.js

Demo:
![RPReplay_Final1591869858](https://user-images.githubusercontent.com/123822/84372737-de100200-abdb-11ea-955c-c6a1928eb220.gif)

RamyRais commented:
Hey, would you be interested to add the ability to render multiple progress bar in one page?
I have this [fork](https://github.com/RamyRais/multi-nprogress) and would be nice to combine them.

zenflow reviewed (approved):
@rstacruz This looks amaze-ballss! Looks like it's done, too. Merge & release it! :D

MickL commented:
Hey @rstacruz , adding Sass and TypeScript types is exactly what I was looking for! Seems like you finished all the work already and just the documentation is missing? Also I couldnt find the TypeScript typings so I needed to additionally install `@types/nprogress` :(

Btw. could you please also make the `z-index`, `top` and `box-shadow` a variable? I have a fixed header so I might need to customize those values.

silasabbott commented:
Just checking up on this—@rstacruz let us know how we can help 🎉

bombillazo commented:
Any update on this package, or has it been abandoned? Would like to know either way what the plan is for this component.
